daemon/images/images.go

```diff
@@ -84,13 +77,18 @@ func (i *ImageService) Images(imageFilters filters.Args, all bool, withExtraAttr
 		return nil, err
 	}
 
-	images := []*types.ImageSummary{}
-	var imagesMap map[*image.Image]*types.ImageSummary
-	var layerRefs map[layer.ChainID]int
-	var allLayers map[layer.ChainID]layer.Layer
-	var allContainers []*container.Container
+	allImages := i.imageStore.Map()
+	selectedImages := allImages
```

Comment:
(As a follow up?) we should probably only get `allImages` if either `danglingOnly` is false and/or if `withExtraAttrs` is set. I see that `i.imageStore.Map() -> `i.imageStore.imagesMap()` -> `NewFromJSON()` will read the image config for every image it includes from disk (and JSON Unmarshal). Perhaps we should avoid that if we're not gonna be needing _all_ images.

Something like

```patch
diff --git a/daemon/images/images.go b/daemon/images/images.go
index 81ea5c5103..882bb3da84 100644
--- a/daemon/images/images.go
+++ b/daemon/images/images.go
@@ -77,10 +77,12 @@ func (i *ImageService) Images(imageFilters filters.Args, all bool, withExtraAttr
                return nil, err
        }

-       allImages := i.imageStore.Map()
-       selectedImages := allImages
+       var allImages, selectedImages map[image.ID]*image.Image
        if danglingOnly {
                selectedImages = i.imageStore.Heads()
+       } else {
+               allImages = i.imageStore.Map()
+               selectedImages = allImages
        }

        var (
@@ -209,6 +211,9 @@ func (i *ImageService) Images(imageFilters filters.Args, all bool, withExtraAttr
                allLayers := i.layerStore.Map()
                layerRefs := make(map[layer.ChainID]int, len(allLayers))
                // Count layer references across all known images
+               if allImages == nil {
+                       allImages = i.imageStore.Map()
+               }
                for _, img := range allImages {
                        rootFS := *img.RootFS
                        rootFS.DiffIDs = nil
```

Reply:
https://github.com/moby/moby/pull/42550/commits/85a37a5a4bc3be6bce4b6e936af5c78a6afb8360